napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,122 @@
+from __future__ import annotations
+
+import logging
+from concurrent.futures import Executor, Future, ThreadPoolExecutor
+from threading import RLock
+from typing import Dict, Iterable, Optional, Tuple
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+LOGGER = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    def __init__(self):
+        self.events = EmitterGroup(source=self, ready=Event)
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self.lock = RLock()
+
+    def slice_layers_async(
+        self, layers: Iterable[Layer], dims: Dims
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer.
+
+        Submitting multiple layers at one generates multiple requests (stored in a dict),
+        but only ONE task.
+
+        A slice submitted with multiple layers will only cancel if the same tuple of layers
+        is reubmitted. If a second slice is requested with only one layer from the first
+        slice, it will not be cancelled.
+        """
+        # Cancel any tasks that are slicing a subset of the layers
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        if existing_task := self._find_existing_task(layers):
+            LOGGER.debug('Cancelling task for %s', layers)
+            existing_task.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            # TODO: temporary to prove a point (which currently fails)
+            # import weakref
+            # assert isinstance(layer, weakref.ReferenceType) # this fails
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # create task for slicing of each request/layer
+        task = self._executor.submit(self._slice_layers, requests)
+
+        # construct dict of layers to layer task
```

Comment:
This seems out of date. Maybe something like:

```suggestion
        # store task for cancellation logic
```